### PR TITLE
automatic formatted media data capability

### PR DESCRIPTION
**Pull request by chrisadubois**
slapped this together in a few hours today to get something in for the share client which needs data asap, we can add to it as we discover more about how we use it.

I tried to use what i had already created in the SDK for history/stats/analysis and that seemed to work to fulfill the requirement.

We could build this in the devices client pretty easily, but it seemed beneficial to add to the SDK to allow other 3rd party to use it since we are going to take the default :)

adds for share,audio,video -> flattens data to desired
automatically gathers based on config value

https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-117100

here's a sample output for share data that we can track on grafana for example
Screen Shot 2020-01-28 at 4 50 41 PM

Please review, thanks !

edit:

sample data printed in grafana
Screen Shot 2020-01-30 at 12 11 44 PM
Screen Shot 2020-01-30 at 12 17 17 PM

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/config.js [hunk 1]
```diff
@@ -25,8 +25,15 @@ export default {
...
     },
     stats: {
+      // 1 second intervals
       interval: 1000,
-      historyMax: 1800
+      // hold the last 2 minute of a calls data
+      historyMax: 120,
+      // should we move these to a separate config key?
+      // send a metrics event automatically
+      autoMediaMetrics: false,
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/config.js]: I think we should rename this to something like `sendDiagnostics` or `autoUploadLogs` to make it more apparent to the end user what it does.

File: packages/node_modules/@webex/plugin-meetings/src/config.js [hunk 2]
```diff
@@ -25,8 +25,15 @@ export default {
...
+      // dispatch data every 10 seconds
+      mediaMetricsInterval: 10
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/config.js]: Why is this different than the other interval? Seems like the naming could be tweaked here, too. Love all the configuration options though.

File: packages/node_modules/@webex/plugin-meetings/src/stats/data.js [hunk 3]
```diff
@@ -13,6 +16,28 @@ export default class WebRTCData {
...
   }
 
+  /**
+   * get flattened rtc/rtcp/rtp/track/transport/candidate data with omitted data (default)
+   * @returns {Object}
+   */
+  flatten() {
+    let flat = {};
+
+    Object.keys(this.data).forEach((key) => {
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/stats/data.js]: Nit: I think `reduce` could accomplish the same thing without rewriting the flat variable in place.

File: packages/node_modules/@webex/plugin-meetings/src/stats/data.js [hunk 4]
```diff
@@ -13,6 +16,28 @@ export default class WebRTCData {
...
+      flat = {...flat, ...omit(this.data[key], DEFAULT_FLATTEN_DATA_KEYS)};
+    });
+
+    return flat;
+  }
+
+  /**
+   * get the unique id for this specific stat pull
+   * @returns {Boolean};
+   */
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/stats/data.js]: Is this comment accurage? uniqueId on a bool?

File: packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+          metrics: true,
+          onData: () => {
+            this.checkData(STATS.audioReceiverId, this.stats.getReceiver(STATS.audioReceiverId));
+          }
+        },
+        {
+          id: STATS.videoReceiverId,
+          correlate: STATS.videoCorrelate,
+          metrics: true,
+          onData: () => {
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js]: I'm guessing we don't control the interval for getStats() so we have to do the check on every event message?

File: packages/node_modules/@webex/plugin-meetings/src/stats/stats.js [hunk 6]
```diff
@@ -155,6 +167,20 @@ export default class MeetingStats extends StatelessWebexPlugin {
     }
   }
 
+  /**
+   * @param {WebRTCData} data
+   * @returns {undefined}
+   * @public
+   * @memberof MeetingStats
+   */
+  doMetrics(data) {
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/stats/stats.js]: DO IT!

Comment (andrewholsted) [reply packages/node_modules/@webex/plugin-meetings/src/stats/stats.js]: nit: `data.data` typically drives me nuts

File: packages/node_modules/@webex/plugin-meetings/src/config.js [hunk 7]
```diff
@@ -25,8 +25,15 @@ export default {
       }
     },
     stats: {
+      // 1 second intervals
       interval: 1000,
-      historyMax: 1800
+      // hold the last 2 minute of a calls data
+      historyMax: 120,
+      // should we move these to a separate config key?
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/config.js]: Please remove this comment.

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 8]
```diff
@@ -293,6 +293,20 @@ export const DEFAULT_TRANSFORM_KEYS = [
   'rtcCandidatePairAudio',
   'rtcCandidatePairVideo'
 ];
+export const DEFAULT_FLATTEN_DATA_KEYS = [
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js]: Would be great to add a comment above this to indicate where these constants come from and are expected to be used.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 9]
```diff
@@ -983,6 +992,34 @@ export default class Meeting extends StatelessWebexPlugin {
     return this.members;
   }
 
+  /**
+   * Reference to the mediaMetrics object
+   * @returns {MediaMetrics}
+   * @public
+   * @memberof Meeting
+   */
+  getMediaMetrics() {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: I know for consistency sake we are using getters around the meetings plugin, but was there a particular reason not to expose just the `this.mediaMetrics` property?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 10]
```diff
@@ -983,6 +992,34 @@ export default class Meeting extends StatelessWebexPlugin {
...
+  /**
+   * Automatically publishes media metrics data flattened and on a config interval
+   * uses the existing getStats data filter, so no new streams are created
+   * but it is separate than history, so history will not be available
+   * @returns {MediaMetrics}
+   * @public
+   * @memberof Meeting
+   */
+  autoStats() {
+    const automatic = new StatsMetrics(this.attrs, this.options);
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: Renaming the var might make this a bit more readable:

```suggestion
    const automaticMetrics = new StatsMetrics(this.attrs, this.options);
```

File: packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+          interval
+        }
+      );
+      metrics.clear();
+    }
+  }
+
+  /**
+   * sets up all the default senders and receivers getStats to collect data
+   * @returns {undefined}
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js]: This is incorrect

File: packages/node_modules/@webex/plugin-meetings/src/stats/stats.js [hunk 12]
```diff
@@ -387,6 +438,10 @@ export default class MeetingStats extends StatelessWebexPlugin {
     return this.history;
   }
 
+  getMetrics() {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/stats/stats.js]: This needs jsdoc

**Comment by chrisadubois** [comment]
@adamweeks we have zero tests around getStats currently -- for this PR to make it in, you'd like to me to add all of the following? unit, integration, and samples e2e tests -- I don't think all of these would be useful and most of them we can't do without media, I'd like to have a discussion on the ones we think are most important to get a mvp in for the feature.

**Comment by arun3528** [comment]
Were you able to find the bandwidth values?. i guess its under the transport

File: packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js (hunk 5 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js]: well, they are separate between getStats and the metrics needed for clients, and can be configured. getStats can be done entirely independently, so short answer yes.

File: packages/node_modules/@webex/plugin-meetings/src/stats/stats.js (hunk 6 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/stats/stats.js]: i also hate it but it got messed up early on and now it's a decent sized refactor...

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 9 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: no particular reason, just consistency sake.

**Comment by chrisadubois** [comment]
@arun3528 yes I found bandwidth values and added them to the report they are on the new candidatepair. so in this PR i also updated to match the newer fresher getStats API

added some unit tests, the getStats information which autoStats pulls from is already tested in the integration test and I don't want to slow it down another bit to wait for autoStats since they would be testing the same thing.

File: packages/node_modules/@webex/plugin-meetings/src/stats/data.js [hunk 13]
```diff
@@ -13,6 +16,28 @@ export default class WebRTCData {
...
+
+    Object.keys(this.data).forEach((key) => {
+      flat[key] = omit(this.data[key], DEFAULT_OMISSION_DATA_KEYS);
+    });
+
+    return flat;
+  }
+
+  /**
+   * get the unique id for this specific stat pull
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/stats/data.js]: I still don't think this returns an id

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 14]
```diff
@@ -764,6 +780,15 @@ export const SDP = {
 };
 
 export const STATS = {
+  audioCorrelate: 'audio',
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/constants.js]: Any reason why these are not all uppercase like the ones below it?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 15]
```diff
@@ -983,6 +992,34 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   */
+  getMediaMetrics() {
+    return this.mediaMetrics;
+  }
+
+  /**
+   * Automatically publishes media metrics data and on a config interval
+   * uses the existing getStats data filter, so no new streams are created
+   * but it is separate than history, so history will not be available
+   * @returns {MediaMetrics}
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: This method doesn't `return`.

File: packages/node_modules/@webex/plugin-meetings/src/stats/stats.js [hunk 16]
```diff
@@ -155,6 +167,20 @@ export default class MeetingStats extends StatelessWebexPlugin {
...
 
+  /**
+   * @param {WebRTCData} data
+   * @returns {undefined}
+   * @public
+   * @memberof MeetingStats
+   */
+  doMetrics(data) {
+    if (this.metrics && data.data) {
+      if (!data.data.isEmpty()) {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/stats/stats.js]: `isEmpty` is a function on the `data.data` property?

File: packages/node_modules/@webex/plugin-meetings/test/integration/spec/journey.js [hunk 17]
```diff
@@ -44,6 +44,12 @@ skipInNode(describe)('plugin-meetings', () => {
       // TODO: end is not a function on browser object
     });
 
+    describe('Check configuration values', () => {
+      it('check config', () => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/test/integration/spec/journey.js]:
```suggestion
      it('properly sets the `'autoSendMediaDiagnostics`' config', () => {
```

File: packages/node_modules/@webex/plugin-meetings/src/stats/transformer.js [hunk 18]
```diff
@@ -10,17 +10,19 @@ StatsTransformer.simplify = (stat) => find(keys(DEFAULT_TRANSFORM_REGEX), (key)
   let value = (DEFAULT_TRANSFORM_REGEX[key].regex.test(stat.id) ? key : false);
 
   if (value) {
-    // some types need to check against a type field as well
-    if (DEFAULT_TRANSFORM_REGEX[value].profiler) {
-      const mapStatToReadableType = stat[DEFAULT_TRANSFORM_REGEX[value].profiler.type];
+    // TODO: wait a year to see if webrtc readds their different implementation for rtp/rtc/transport stats
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/stats/transformer.js]: Is there a link we could add to this comment to track the work being done on this, or the limitation?

**Comment by adamweeks** [comment]
The merge commit needs to be replaced by a rebase on master and the tests are failing.

**Comment by adamweeks** [comment]
Closing for now, @chrisadubois will reopen when ready.

**Pull request closed without merging**